src/pip/_internal/resolution/resolvelib/factory.py

```diff
@@ -138,15 +158,19 @@ def iter_found_candidates(self, ireq, extras):
         # Return installed distribution if it matches the specifier. This is
         # done last so the resolver will prefer it over downloading links.
         if (installed_dist is not None and
+                not can_upgrade and
                 installed_dist.parsed_version in ireq.req.specifier):
             yield self._make_candidate_from_dist(
                 dist=installed_dist,
                 extras=extras,
                 parent=ireq,
             )
 
-    def make_requirement_from_install_req(self, ireq):
-        # type: (InstallRequirement) -> Requirement
+    def make_requirement_from_install_req(self, ireq, root=False):
+        # type: (InstallRequirement, bool) -> Requirement
+        if root and ireq.name:
+            self.root_reqs.add(ireq.name)
```

Comment:
What’s the difference between `ireq.is_direct` and this additional argument? Also my gut feeling is we need `canonicalize_name()` here.

Reply:
Possibly nothing... I went with an explicit `root` argument just because my thought process was "we process the root requirements first - those are the ones we want to upgrade, so cool". I'm also always a little wary of relying on requirement attributes - we clone them a lot and I worry we'll lose something. But you're right, I'll rework it to use `is_direct` and see how that goes.

I'll track the question of `canonicalize_name` back too. Maybe I can check this in the test by capitalising the project name. It would be good to explicitly test for it, as it's so easy to miss.